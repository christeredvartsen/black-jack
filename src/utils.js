const gameState = {
  BOTH_WINS: 'both win',
  BOTH_LOSE: 'both lose',
  P1_WINS: 'player 1 wins',
  P2_WINS: 'player 2 wins',
  P1_DRAWS: 'player 1 to draw',
  P2_DRAWS: 'player 2 to draw',
};

const getCardValue = ({ value }) => {
  let result;

  switch (value.toLowerCase()) {
    case 'j': // continue
    case 'q': // continue
    case 'k': result = 10; break;
    case 'a': result = 11; break;
    default: result = parseInt(value, 10);
  };

  return result;
};

const getCardsScore = (cards) => cards.reduce((score, card) => score + getCardValue(card), 0);

const getGameState = (p1, p2) => {
  if (p1 === 21 && p2 === 21) {
    return gameState.BOTH_WINS;
  } else if (p1 > 21 && p2 > 21) {
    return gameState.BOTH_LOSE;
  } else if (p2 > 21 || p1 === 21) {
    return gameState.P1_WINS;
  } else if (p1 > 21 || (p1 >= 17 && p2 > p1)) {
    return gameState.P2_WINS;
  } else if (p1 < 17) {
    return gameState.P1_DRAWS;
  } else {
    return gameState.P2_DRAWS;
  }
};

export {
  getCardValue,
  getCardsScore,
  gameState,
  getGameState,
};