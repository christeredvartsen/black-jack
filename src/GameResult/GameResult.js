import React from 'react';
import PropTypes from 'prop-types';
import { gameState as s } from '../utils';

const GameResult = ({ result, p1, p2 }) => {
  switch (result) {
    case s.BOTH_WINS: return <h2>Both won!</h2>;
    case s.BOTH_LOSE: return <h2>Both lose!</h2>;
    case s.P1_WINS: return <h2>{p1} won!</h2>;
    case s.P2_WINS: return <h2>{p2} won!</h2>;
    default: return null;
  }
}
GameResult.propTypes = {
  result: PropTypes.string.isRequired,
  p1: PropTypes.string.isRequired,
  p2: PropTypes.string.isRequired,
};

export default GameResult;