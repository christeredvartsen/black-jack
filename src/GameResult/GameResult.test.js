import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import each from 'jest-each';
import GameResult from './GameResult';
import { gameState as s } from '../utils';

describe('<GameResult />', () => {
  each([
    [
      s.BOTH_WINS,
      'name 1',
      'name 2',
      'Both won!'
    ],
    [
      s.BOTH_LOSE,
      'name 1',
      'name 2',
      'Both lose!'
    ],
    [
      s.P1_WINS,
      'name 1',
      'name 2',
      'name 1 won!'
    ],
    [
      s.P2_WINS,
      'name 1',
      'name 2',
      'name 2 won!'
    ],
  ]).test('renders as expected', (result, p1, p2, expectedText) => {
    expect(
      shallow(
        <GameResult result={result} p1={p1} p2={p2} />
      ).find('h2').text()
    ).toBe(expectedText);
  });

  test('does not render for all results', () => {
    expect(
      shallow(
        <GameResult result={s.P1_DRAWS} p1="name 1" p2="name 2" />
      ).find('h2').exists()
    ).toBe(false);
  });

  test('matches snapshot', () => {
    const tree = shallow(
      <GameResult result={s.BOTH_WINS} p1="name1" p2="name2" />
    );
    expect(toJson(tree)).toMatchSnapshot();
  });
});