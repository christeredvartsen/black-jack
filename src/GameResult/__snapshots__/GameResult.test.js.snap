// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GameResult /> matches snapshot 1`] = `
<h2>
  Both won!
</h2>
`;
